File: scripts/catchment_extractor.py
# -*- coding: utf-8 -*-
"""
Created on Wed Nov 20 10:54:39 2019

@author: ATeklesadik
"""
import geopandas as gpd
import fiona
import descartes
from osgeo import ogr
from shapely.geometry import Polygon
import numpy as np
################################# run the code for sample admin 
def add_to_list(items,list_):
    for elments in items:
        if elments not in list_:
            list_.append(elments)
    return list_

def fun2(item,basin):
    for i in item:
        item2 = basin[basin['NEXT_DOWN'] == i]['HYBAS_ID'].values
    return item2
        

eth_admin3 = gpd.read_file('C:/documents/ethiopia/admin3/admin3.shp')
basin = gpd.read_file('C:/documents/General_data/Basins/hydrosheds/African_basins/hybas_lake_af_lev12_v1c.shp')
basin = gpd.overlay(basin,eth_admin3, how='identity')




rivers =gpd.read_file('C:/documents/General_data/Basins/hydrosheds/African_rivers/af_riv_15s.shp')
rivers = gpd.overlay(eth_admin3, rivers, how='intersection')

sample_admin = eth_admin3[0:1]

basin_check = gpd.overlay(basin,sample_admin,  how='intersection')
basin_check2 = gpd.overlay(basin,sample_admin, how='identity')

datalist={}
           
def catchment_extractor(sample_admin):
    #basin_check = gpd.overlay(sample_admin, basin, how='intersection')
    basin_check = basin_check2[basin_check2['Pcode'].isin(sample_admin['Pcode'].values)]
    basin_check = basin_check.sort_values(by=['UP_AREA'],ascending=False)
    basin_check1 = basin_check
    basin_check1.drop_duplicates(subset ="HYBAS_ID", keep = 'first', inplace = True) 
    basin_check1 = basin_check1.iloc[:1]
    for j in basin_check1['HYBAS_ID'].values:
        list_con=[]
        list_item=[]
        list_con.append(j)
        item = np.unique(basin[basin['NEXT_DOWN'] == j]['HYBAS_ID'].values)
        print(item)
        list_con=add_to_list(item,list_con)
        ii=100
        #if len(item) >0:
        while ii > 0: #item != []:
            ii=ii-1
            print(ii)
            l1=[]
            if len(item) >0:
                for i in item:
                    itm2 = basin[basin['NEXT_DOWN'] == i]['HYBAS_ID'].values
                    l1.append(len(itm2))
                    list_item.append(itm2.tolist())
                    list_con=add_to_list(itm2,list_con)
                #l2=max(l1)
                item=list_con #list(set(list_item)-set(item.tolist()))
        cachment= basin[basin['HYBAS_ID'].isin(list_con)] 
        datalist[j] = cachment            
    return(datalist)
event_tc=catchment_extractor(sample_admin)


for key,valu in event_tc.items():
    print(key)
    valu.plot()

File: scripts/catchment_extractor.R
library(dplyr)
library(tidyr)
library(purrr)
library(stringr)
library(tmap)
library(maps)
library(httr)
library(sf)
library(lubridate)
library(zoo)
library(xts)
library(lubridate)
require(data.table)

##############################################
catchment_extractor_old<-function(sample_admin){
  basin <- st_read(dsn='C:/documents/General_data/Basins/hydrosheds/African_basins',layer='hybas_lake_af_lev12_v1c')
  basin_check <- st_intersection(sample_admin, basin) %>% arrange(desc(UP_AREA)) %>% dplyr::select(HYBAS_ID,geometry)
  basin_<-basin %>% dplyr::select(HYBAS_ID,geometry)
  datalist = list()
  
  for (j in 1:length(as.list( basin_check$HYBAS_ID)))  { 
    i<- as.list( basin_check$HYBAS_ID)[[j]]
    list_con<-c(i)
    item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID 
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID   #1
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID  #2
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID #4
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID #5
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID #6
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID #7
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID #8
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID #9
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID #10
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID #11
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID #12
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID #13
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID #14
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID #15
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    if(length(item) != 0) { for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID 
    list_con<-c( list_con , item[ ! item %in% list_con ] )    }}      }}      }}      }}      }}      }}      }}      }}      }}      }}      }}      }}      }}      }}     }     }

    cachment<-assign(paste("cat", as.character(j), sep = "_"), basin_[basin_$HYBAS_ID %in% list_con,])
    
    dat <- data.frame(HYBAS_ID = j, geom = cachment$geometry)
    df = st_as_sf(dat)
    datalist[[j]] <- df # add it to your list
      }
  return(datalist)
  
}
##########################################################

catchment_extractor<-function(sample_admin){
  basin <- st_read(dsn='C:/documents/General_data/Basins/hydrosheds/African_basins',layer='hybas_lake_af_lev12_v1c')
  basin_check <- st_intersection(sample_admin, basin) %>% arrange(desc(UP_AREA)) %>% dplyr::select(HYBAS_ID,geometry)
  basin_<-basin %>% dplyr::select(HYBAS_ID,geometry)
  datalist = list()
  
  for (j in 1:length(as.list( basin_check$HYBAS_ID)))  { 
    i<- as.list( basin_check$HYBAS_ID)[[j]]
    list_con<-c(i)
    item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID 
    list_con<-c( list_con , item[ ! item %in% list_con ] )
    while (length(item) != 0)
    {
      for (i in item){item<-basin[basin$NEXT_DOWN==i,]$HYBAS_ID   #1
      list_con<-c( list_con , item[ ! item %in% list_con ] )}
      
    }
    #assign(paste("cat", as.character(j), sep = "_"), list_con)
    cachment<-assign(paste("cat", as.character(j), sep = "_"), basin_[basin_$HYBAS_ID %in% list_con,])
    
    dat <- data.frame(HYBAS_ID = j, geom = cachment$geometry)
    df = st_as_sf(dat)
    datalist[[j]] <- df # add it to your list
    
    
  }
  return(datalist)
  
}

#---------------------- Load admin boundary data -------------------------------  

eth_admin3 <- st_read(dsn='C:/documents/ethiopia/admin3',layer='admin3')
rivers <- st_read(dsn='C:/documents/General_data/Basins/hydrosheds/African_rivers',layer='af_riv_15s')

eth_admin3<-eth_admin3 %>%  dplyr::mutate(Pcode=NewPCODE) %>%  dplyr::select(Pcode,geometry)

################################# run the code for sample admin 
sample_admin<-eth_admin3[1,]

cat_admin1<-catchment_extractor(sample_admin)


 
